File: src/common/common/main/java/settingdust/mod_sets/data/ModSetsCommonConfig.java
package settingdust.mod_sets.data;

import settingdust.mod_sets.util.LoaderAdapter;
import settingdust.mod_sets.util.SavingData;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

public class ModSetsCommonConfig implements SavingData {
    private static final Path configPath = LoaderAdapter.get().getConfigDirectory().resolve("common.json");
    public static Data data = new Data(true);

    @Override
    public void load() {
        try {Files.createDirectories(configPath.getParent());} catch (IOException ignored) {}
        try {
            Files.createFile(configPath);
            Files.writeString(configPath, "{}");
        } catch (IOException ignored) {}
        try {
            data = GSON.fromJson(Files.newBufferedReader(configPath), Data.class);
        } catch (IOException e) {
            throw new RuntimeException("Fail to load common config", e);
        }
        save();
    }

    @Override
    public void save() {
        try {
            GSON.toJson(data, Files.newBufferedWriter(configPath));
        } catch (IOException e) {
            throw new RuntimeException("Fail to save common config", e);
        }
    }

    public record Data(
        boolean badgeInModMenu
    ) {}
}

File: src/common/main/kotlin/settingdust/mod_sets/data/ModSetsCommonConfig.kt
package settingdust.mod_sets.data

import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.decodeFromStream
import kotlinx.serialization.json.encodeToStream
import settingdust.mod_sets.ModSets
import settingdust.mod_sets.data.ModSetsDisabledMods.disabledMods
import settingdust.mod_sets.util.LoaderAdapter
import kotlin.io.path.createDirectories
import kotlin.io.path.createFile
import kotlin.io.path.div
import kotlin.io.path.inputStream
import kotlin.io.path.outputStream
import kotlin.io.path.writeText

@Serializable
data class ModSetsCommonConfig(
    val badgeInModMenu: Boolean = true
) {
    companion object : SavingData {
        var instance: ModSetsCommonConfig = ModSetsCommonConfig()
            private set
        private val configPath = LoaderAdapter.configPath / "common.json"

        @OptIn(ExperimentalSerializationApi::class)
        override fun reload() {
            runCatching { LoaderAdapter.configPath.createDirectories() }
            runCatching {
                configPath.createFile()
                configPath.writeText("{}")
            }
            instance = ModSets.configJson.decodeFromStream(configPath.inputStream())
            save()
        }

        @OptIn(ExperimentalSerializationApi::class)
        override fun save() {
            ModSets.configJson.encodeToStream(disabledMods, configPath.outputStream())
        }
    }
}
